### PR TITLE
Checkpointing source-mssql

### PR DESCRIPTION
**Pull request by rodireich**
## What
*Describe what the change is solving*
*It helps to add screenshots if it affects the frontend.*

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
*Are there any breaking changes? What is the end result perceived by the user?*

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/resources/pull-requests-handbook)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:generateScaffolds` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

<details><summary><strong>Updating the Python CDK</strong></summary>

### Airbyter

Before merging:
- Pull Request description explains what problem it is solving
- Code change is unit tested
- Build and my-py check pass
- Smoke test the change on at least one affected connector
   - On Github: Run [this workflow](https://github.com/airbytehq/airbyte/actions/workflows/connectors_tests.yml), passing `--use-local-cdk --name=source-<connector>` as options
   - Locally: `airbyte-ci connectors --use-local-cdk --name=source-<connector> test`
- PR is reviewed and approved
      
After merging:
- [Publish the CDK](https://github.com/airbytehq/airbyte/actions/workflows/publish-cdk-command-manually.yml)
   - The CDK does not follow proper semantic versioning. Choose minor if this the change has significant user impact or is a breaking change. Choose patch otherwise.
   - Write a thoughtful changelog message so we know what was updated.
- Merge the platform PR that was auto-created for updating the Connector Builder's CDK version
   - This step is optional if the change does not affect the connector builder or declarative connectors.

</details>

### CODE REVIEW

**Comment by rodireich** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34182?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

<b>In this stack:</b> <code>🔍 Enhance Data Snapshotting and Testing</code>
* `master`
  * **#33837** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33837?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **#33848** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33848?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
      * **#33838** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33838?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
        * **#33851** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33851?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
          * **#34063** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34063?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
            * **#34120** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34120?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
              * **#34182** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/34182?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @rodireich and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/mssql/MssqlDebeziumStateUtil.java
Comment (rodireich) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/mssql/MssqlDebeziumStateUtil.java:62]: @xiaohansong as far as I can tell, whether or not there was an event, closing the engine saves a state.

File: airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/CdcMssqlSourceTest.java
Comment (xiaohansong) [reply airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/CdcMssqlSourceTest.java]: fixed!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java
Comment (rodireich) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java:255]: @xiaohansong I don't think `getSourceDefinedPrimaryKey()` is correct here, is it?

File: airbyte-cdk/java/airbyte-cdk/README.md
Comment (akashkulk) [inline airbyte-cdk/java/airbyte-cdk/README.md:169]: not a breaking change per se - but a significant version update. 
Should we make it a 1.0.0 release or let it bake for a bit?

File: airbyte-cdk/java/airbyte-cdk/core/src/main/resources/version.properties
Comment (akashkulk) [inline airbyte-cdk/java/airbyte-cdk/core/src/main/resources/version.properties:1]: 0.17.0

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/CdcTargetPosition.java
Comment (akashkulk) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/CdcTargetPosition.java:77]: Why this change?

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumStateDecoratingIterator.java
Comment (akashkulk) [inline airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/internals/DebeziumStateDecoratingIterator.java:53]: nit : here and below, new variable on new line

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlCdcHelper.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlCdcHelper.java:98]: Side not : Opened this issue https://github.com/airbytehq/airbyte/issues/34812 to just get rid of heartbeats altogether

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java:45]: Ok if not handled in this PR : 

As we move towards a world where the initial snapshot code supports a resumable full refresh, should we do away with the "end" position queries? This was added in MySQL to prevent the case where a rapidly growing database would never have a full snapshot completed because records are added faster than we can read. In practice, I'm not sure this is an issue. 

So basically, what do you think of having repeated queries such as : 
`select * from table where pk >= previous_pk_val limit CHUNK_SIZE`

instead of 
`select * from table where pk >= previous_pk_val and pk < max_pk limit CHUNK_SIZE`

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java:44]: Why do we need legacy state again?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java:75]: can we hardcode this in a centralized place?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialLoadRecordIterator.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialLoadRecordIterator.java:119]: not : String.format("SELECT...>", wrappedColumnNames, fullTableName, etc)

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialLoadRecordIterator.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialLoadRecordIterator.java:106]: what does OC stand for?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java
Comment (akashkulk) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java:115]: can we just throw an exception if the saved offset is no longer present? This is line with new behavior telling the user to reset their streams

**Review by akashkulk — APPROVED** [review]
Some minor nits and open questions - but nothing blocking getting this in. Great work, Rodi!

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/CdcTargetPosition.java
Comment (rodireich) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/CdcTargetPosition.java:77]: This came with Marius' heartbeat change
@postamar Why did the default change?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java:44]: This was lifted from MySqlCursorBasedStateManager.
The idea I think is to not break existing saved legacy state already in db

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialLoadRecordIterator.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialLoadRecordIterator.java:106]: OrderedColumn (the differentiate from PK)

File: airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/CdcTargetPosition.java
Comment (postamar) [reply airbyte-cdk/java/airbyte-cdk/db-sources/src/main/java/io/airbyte/cdk/integrations/debezium/CdcTargetPosition.java:77]: A default value of `true` makes checkpointing impossible, see implementation of `DebeziumStateDecoratingIterator::computeNext`: `checkpointOffsetToSend` will never be assigned a value. This is despite the fact that the purpose of the test involving `isSameOffset`, which guards this assignment, appears to be to avoid emitting more than one state message with the same state.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/initialsync/MssqlInitialReadUtil.java:115]: I'll take this offline - need to think since user will see more failures as opposed to connector handling this case quietly.
To compare: postgres and mysql are still going to fall back to initial and not throw an error.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java:45]: Is this necessary? 
Since the first incremental sync is now queued immediately after initial completes, any record whose `pk > max_pk` is going to be included in that incremental sync.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/cursor_based/MssqlCursorBasedStateManager.java]: Done

**Review by lazebnyi — APPROVED** [review]

**Comment by edgao** [comment]
@rodireich - can you hold off on merging this pr for a little bit? @subodh1810 published cdk 0.17.1 [here](https://github.com/airbytehq/airbyte/pull/35027)

**Review by postamar — COMMENTED** [review]
No major complaints!

**Comment by rodireich** [comment]
No problem @edgao 
I already sent for publishing cdk v0.18
not sure whose got in first

**Comment by subodh1810** [comment]
@rodireich I think you will have to republish the CDK after pulling in my changes! I merged the PR so you can pull in master and re-publish

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java:87]: Since this is informational log I never want it to stop the sync.
There are some variations that are not known - all sorts of hosted and managed and serverless db services won't let you look under the hood.
In the past we ended up adding this to equivalent pre sync logs - getting the version and engine of mysql for example.
Not sure what you meant with `final`

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java:94]: This max value is a must to be able to upper bound our sync - so failing to retrieve it should error out.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java
Comment (rodireich) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlQueryUtils.java:137]: I looked for and couldn't find.
Implementation will have the other way around typically byte size to human readable:
[FileUtils.byteCountToDisplaySize](https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FileUtils.html#byteCountToDisplaySize(java.math.BigInteger) for example

**Review by octavia-approvington — APPROVED** [review]

**Comment by octavia-approvington** [comment]
Lets merge it <br/> sheep thumbs up

**Pull request merged**
